### ISSUE TITLE
Exception in thread "main" java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Float

### ISSUE THREAD

**Issue by CallMeByYourName** [openXBOW/openXBOW#4]
I have 1023 LLDs for each audio and when I enter the LLDS file and the label file, I get the error 

Parsing input ...
Reorganizing input data ...
Creating codebook ...
Exception in thread "main" java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Float
        at openxbow.codebooks.CodebookNumericTrainingSelector.getTrainingFeatures(CodebookNumericTrainingSelector.java:162)
        at openxbow.codebooks.CodebookNumericTrainingSelector.select(CodebookNumericTrainingSelector.java:70)
        at openxbow.codebooks.CodebookNumericTrainingSelector.<init>(CodebookNumericTrainingSelector.java:58)
        at openxbow.codebooks.CodebookNumericTrainingSelector.<init>(CodebookNumericTrainingSelector.java:50)
        at openxbow.codebooks.HyperCodebook.generateCodebook(HyperCodebook.java:85)
        at openxbow.main.OpenXBOW.main(OpenXBOW.java:97)

Some of the contents of the llds file are as follows:

train_0001;-1.58E+00;-1.44E+00;-2.20E+00;-1.30E+00;-2.02E+00;-2.21E+00;-2.19E+00;-1.16E+00;-1.72E+00;-2.05E+00;-1.99E+00
train_0001;-2.33E+00;-2.13E+00;-2.32E+00;-2.11E+00;-1.04E+00;-1.51E+00;-1.83E+00;-1.65E+00;-2.13E+00;-1.99E+00;-1.98E+00
train_0001;-2.02E+00;-2.32E+00;-2.34E+00;-2.04E+00;-2.31E+00;-2.29E+00;-2.37E+00;-2.06E+00;-2.17E+00;-8.96E-01;-1.54E+00
train_0001;-1.51E+00;-1.51E+00;-1.88E+00;-1.78E+00;-1.62E+00;-1.69E+00;-2.17E+00;-2.10E+00;-1.94E+00;-2.04E+00;-1.96E+00
train_0001;-2.00E+00;-2.04E+00;-2.00E+00;-2.42E+00;-2.25E+00;-2.34E+00;-2.34E+00;-1.99E+00;-2.10E+00;-2.28E+00;-2.33E+00

Some of the contents of the label file are as follows:

name;label
train_0001;V
train_0002;V
train_0003;E
train_0004;V
train_0005;E
train_0006;V
train_0007;E
train_0008;O
train_0009;V
train_0010;V
train_0011;O
train_0012;V
train_0013;O

The code I entered is:

java -jar openXBOW.jar -i llds.CSV  -l label_new.CSV -o bow.csv

Can you help me to solve this problem?

**Comment by maxschmitt** [comment]
Sure, I see two problems here:

1) openXBOW does not support uppercase file endings, i.e., you should use `llds.csv` and `label_new.csv`. However, based on the error reported by you, this seems not to be your issue (maybe you resolved this already).

2) Given the Exception thrown, I think that there is some issue with your `llds` file. 
Based on the excerpts you reported, I cannot reproduce the error, but it seems that there is a string or character in one cell of the csv file. Please cross-check that there are only numbers (except for the first column with the file name) in your file.
Of course, also symbolic (text) features are supported, but then, a symbolic feature must be present consistently across all rows of your file.

**Comment by CallMeByYourName** [comment]
Thank you for your reply.

I learned about openXBOW in "A Bag of Wavelet Features for Snore Sound Classification". And I want to reproduce the method.

By writing the .m file through MATLAB R2020b I obtained 1023 LLDs based on wavelet packets.

I have encountered some problems:

   1.Using method "writetable" I can only get csv files with "," as separator, I wrote separate python code to replace the separator 
      with ";" (maybe that's what went wrong).
   2.Does llds.csv in openXBOW support a row with 1023 columns of LLDs, because I noticed that the llds.csv in example has line 
      feed handling. So I divided the 1023 llds into 93 rows of 11 llds each.

I'm sorry if my questions have caused you any trouble.

**Comment by maxschmitt** [comment]
No worries, thanks for your interest in our research and the toolkit.

1. Separator "," in csv files should be handled automatically by openXBOW (as long it is consistent across the file). So, you could try it also with the original file.

2. 1023 columns should not be an issue. Dividing llds into several rows will result in the issue that only 11 llds are taken into account in total (the number of columns) and 93 different llds will be aggregated into 1 "lld", which might not make sense. Thus, I would suggest in any case that you keep all 1023 different llds from one time frame in one row.

Please let me know if the problems persist.

**Comment by CallMeByYourName** [comment]
Sorry, I still can't solve the problem.

In MATLAB, my llds are formatted as 
     data[828*1023 double] 
and my filename format is 
     sample[828*1 string]
class
     class[828*1 char]

I use the code:

     >>llds = table(sample, data);
     >>writetable(llds,'llds.csv','Delimiter',';');

     >>various = {'name', 'label'};
     >>labels = table(sample, class, 'VariableNames', various);
     >>writetable(labels, 'labels.csv', 'Delimiter', ';');

to generate llds.csv and labels.csv.

Since the first line of the generated llds.csv file is the name of the column, I used Notepad to delete this line.

As a result, the same error is still reported when running the program.

**Comment by maxschmitt** [comment]
I don't see any issue in your code, but I cannot reproduce it, as I don't have Matlab.

If you sent me some more samples from your llds and labels files, I could check if I run into the same exception.

**Comment by CallMeByYourName** [comment]
Here is my llds.csv (I did not delete the column information in the first row) and labels.csv.

https://github.com/CallMeByYourName/SnS

Thank you very much for your help!

**Comment by maxschmitt** [comment]
I found an issue with the LLDs for "train_0043": In the column "WPTEFmat_937", there is written "-Inf", which is not supported, as no distances can be computed with infinity. Thus, you need to handle such cases after the feature extraction (maybe there are some more "-Inf" entries, I have not checked this).

Furthermore, when comparing with the approach from the paper you cited, the WT features were extracted from short frames of a length between 16 ms and 64 ms there. By this, you will obtain a number of LLD rows for each instance (with the same instance name), which are then aggregated as one BoAW feature vector for each file by openXBOW.

**Comment by CallMeByYourName** [comment]
I see, it should be a problem of data errors generated in the generation of features. I will modify the code to fix this issue.

You are right, I am currently extracting llds in 32ms frames. I am currently extracting the WPT features.

There are a few points about WT that I'm not sure about.

1. I used Matlab's wavedec function to perform wavelet decomposition, and then used the wenergy function to get Ea [1*1 double] (the percentage of energy corresponding to the approximation) and Ed [1*3 double] (the vector containing the percentages of energy corresponding to the details). If I split a piece of audio into 175 frames, I will get 4*175 parameters.

Next, four llds need to be calculated

-mean, the average of all Ea and the average of all Ed. [1*4 double]
-variance, the variance of all Ea and the variance of all Ed. [1*4 double]
-waveform length (the sum of the absolute differences)
This one is a little hard for me to understand. I am adding up all the differences after calculating the absolute difference 
between two adjacent numbers in the vector. Like this

>sumEa = 0
>for i = 2:175
>tmp = abs(Ea(i) - Ea(i-1))
>sumEa = sumEa + tmp
>end

I don't know if that's right, and finally got [1*4 double].
-entropy, I considered using either Shannon entropy or log entropy, and ended up using Shannon entropy. [1*4 double].

This gives 16 llds for an audio when the maximum number of layers of wavelet decomposition is 3 as stated in the paper.

Can you tell me if I'm doing it the right way?

2. WPTE
sss

**Comment by CallMeByYourName** [comment]
I see, it should be a problem of data errors generated in the generation of features. I will modify the code to fix this issue.

You are right, I am currently extracting llds in 32ms frames. I am currently extracting the WPT features.

There are a few points about WT that I'm not sure about.

1. I used Matlab's wavedec function to perform wavelet decomposition, and then used the wenergy function to get Ea [1*1 double] (the percentage of energy corresponding to the approximation) and Ed [1*3 double] (the vector containing the percentages of energy corresponding to the details). If I split a piece of audio into 175 frames, I will get 4*175 parameters.
Next, four llds need to be calculated
-mean, the average of all Ea and the average of all Ed. [1*4 double]
-variance, the variance of all Ea and the variance of all Ed. [1*4 double]
-waveform length (the sum of the absolute differences)
This one is a little hard for me to understand. I am adding up all the differences after calculating the absolute difference
between two adjacent numbers in the vector. Like this
> sumEa = 0
for i = 2:175
tmp = abs(Ea(i) - Ea(i-1))
sumEa = sumEa + tmp
end
>
I don't know if that's right, and finally got [1*4 double] parameters.
-entropy, I considered using either Shannon entropy or log entropy, and ended up using Shannon entropy. [1*4 double].
This gives 16 llds for an audio when the maximum level of layers of wavelet decomposition is 3 as stated in the paper.

2. WPTE
Since wenergy function can only calculate the percentages of energy corresponding to the terminal nodes of the wavelet packet tree T. And as mentioned in the paper, the coefficients of each node of each layer are needed.
Therefore I use the wpdec function for wavelet packet decomposition and then use wpcoef function to obtain the coefficients of each layer. Follow the formula to calculate
![image](https://user-images.githubusercontent.com/45512150/115653699-34a0ad80-a362-11eb-95ae-a703e76649a2.png)
The calculated should be obtained as llds, and when the maximum level of layers of wavelet packet decomposition is 9, 1023 llds are obtained.

Could you please help me to see if my calculation is wrong?

**Comment by CallMeByYourName** [comment]
I see, it should be a problem of data errors generated in the generation of features. I will modify the code to fix this issue.
You are right, I am currently extracting llds in 32ms frames. I am currently extracting the WPT features.
There are a few points about WT that I'm not sure about.

1.WT features (16ms per frame)
Suppose an audio file is divided into 175 frames. 

For each frame, I first perform a level 3 wavelet decomposition of 'bior2.8' using the _wavedec_ function. Three vectors of approximation coefficients and three vectors of detail coefficients can be obtained by the _appcoef_ function and _detcoef_ function.
>（A1[136 * 1 double],A2 [76 * 1 double],A3 [46 * 1 double],D1 [136 * 1 double],D2 [76 * 1 double],D3 [46 * 1 double]）
>
According to the Eq
![image](https://user-images.githubusercontent.com/45512150/115668159-548f9b80-a379-11eb-95a3-b0701e00e4e1.png)
**Here I have a question about , is it the percentage of the square of one element of a vector in the sum of the squares of all elements of the vector of all decomposition levels?**
6 vectors can be calculated. The calculation method I use is
>Ea(i) = （Ai.^2) / [sum(A1.^2)+sum(A2.^2)+sum(A3.^2)] * 100, i = 1,2,3
Ed(i) = （Di.^2) /  [sum(D1.^2)+sum(D2.^2)+sum(D3.^2)] * 100, i = 1,2,3
>

According to the paper, use all three approximation coefficients famalies , plus one family of detail coefficients in the first level.
>Ed1 [136 * 1 double], Ea1 [136 * 1 double], Ea2 [76 * 1 double], Ea3 [46 * 1 double].
>
Calculate _mean_, _variance_, _waveform length_, and _entropy_ separately for these four vectors.

·_mean_, calculate the average value of the elements in the vector.
·_variance_, calculate the variance of all elements in the vector.
·_waveform length_, the sum of the absolute differences, this one is a little hard for me to understand. I am adding up all the differences after calculating the absolute difference between two adjacent numbers in the vector. Like this
>sumEa1 = 0
for i = 2:136
tmp = abs(Ea1(i) - Ea1(i-1))
sumEa1 = sumEa1 + tmp
end
>
·_entropy_, **I considered using either Shannon entropy or log entropy, and ended up using Shannon entropy.**
>wentropy(Ea1, 'shannon');
>
This way I can generate 16 LLDs per frame (is this one BoAW feature vector?). One audio file generates 175 * 16 LLDs.
So for an audio file the input to openXBOW is a 175 row, 16 column matrix?

2.WPTE (32ms per frame)
Suppose an audio file is divided into 131 frames. 
For each frame, a 9-layer wavelet packet decomposition of 'haar' is performed using the _wpdec_ function, and then use _wpcoef_ function to obtain the coefficients of each subband. For each subband, using the Eq
![image](https://user-images.githubusercontent.com/45512150/115672889-95d67a00-a37e-11eb-85b3-256174bf67ef.png)
For the 9-layer wavelet decomposition, there are 1023 subbands in total. A value can be computed for each subband so that each frame yields a vector of length 1023.
Should the input be a matrix with 131 rows and 1023 columns?

Could you please help me to see if my calculation is wrong?

**Comment by maxschmitt** [comment]
Your assumption concerning the shape is correct. However, the number of rows (=frames) generated for one audio file can be variable, depending on the duration of the audio.

Concerning the percentage of the square and the entropy, I am not very sure. Maybe you would need to address your question to the first author of the paper who was doing the feature extraction.

**Comment by CallMeByYourName** [comment]
I'm clear on this.

Again, thank you for your help, it is really important for my current research. 

:)

**Comment by maxschmitt** [comment]
No worries, let me know in case you are facing further issues. ;)

**Comment by CallMeByYourName** [comment]
Hello, I would like to ask, if a llds.csv complete input and split into several csv inputs, **codebook size** and **number of assignments** are the same, will the output result be different?

**Comment by maxschmitt** [comment]
Well, the output will differ if you train the codebook on llds.csv. The codebook can be trained only once on a fixed set of LLDs (i.e., on one file). Otherwise, you would have several codebooks. (This does not apply for the "generic" and the "pdf" options to generate codebooks, where codebooks are data-independent and will be exactly the same, independent from the actual training data.)

When you use a fixed codebook, however, the outputs will be the same, no matter whether you split the input files or not.

**Comment by CallMeByYourName** [comment]
Okay, so it looks like I need to do a BoAW extraction of llds.csv for a training set and save the codebook, and then use that codebook directly for the test and development sets.

Sorry, it was me who didn't read the README.md file carefully, which explains how to store and load codebooks.

**Comment by maxschmitt** [comment]
Yes, that's how it works, no worries.
However, in case your training set is too large, for the learning of the codebook, you can also consider a reduced (subsampled) version of your training llds.csv, as the performance of the BoAW features is not much affected by the amount of LLDs, where the codebook is trained from.
